From b16a45a39f1e4139e897b66801fc707d56201fff Mon Sep 17 00:00:00 2001
From: Chad Dupuis <cdupuis@redhat.com>
Date: Thu, 21 Aug 2014 13:05:49 -0400
Subject: [ethernet] qlcnic: Enable VF flood bit on PF

Message-id: <1408626422-3941-66-git-send-email-cdupuis@redhat.com>
Patchwork-id: 88602
O-Subject: [PATCH RHEL 7.1 065/138] qlcnic: Enable VF flood bit on PF.
Bugzilla: 1089352
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Sucheta Chakraborty <sucheta.chakraborty@qlogic.com>

Bug 1089352 (https://bugzilla.redhat.com/show_bug.cgi?id=1089352)

net commit id 820b52fd80dfca80a1fa111f9896868662e58791

o On enabling VF flood bit, PF driver will  be able to receive traffic
  from all its VFs.

Signed-off-by: Sucheta Chakraborty <sucheta.chakraborty@qlogic.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_sriov_pf.c b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_sriov_pf.c
index 1bd0c06..cce9e14 100644
--- a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_sriov_pf.c
+++ b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_sriov_pf.c
@@ -15,6 +15,8 @@
 #define QLC_MAC_OPCODE_MASK 0x7
 #define QLC_MAC_STAR_ADD 6
 #define QLC_MAC_STAR_DEL 7
+#define QLC_VF_FLOOD_BIT BIT_16
+#define QLC_FLOOD_MODE  0x5
 
 static int qlcnic_sriov_pf_get_vport_handle(struct qlcnic_adapter *, u8);
 
@@ -340,6 +342,28 @@ static int qlcnic_sriov_pf_cfg_vlan_filtering(struct qlcnic_adapter *adapter,
  return err;
 }
 
+/* On configuring VF flood bit, PFD will receive traffic from all VFs */
+static int qlcnic_sriov_pf_cfg_flood(struct qlcnic_adapter *adapter)
+{
+ struct qlcnic_cmd_args cmd;
+ int err;
+
+ err = qlcnic_alloc_mbx_args(&cmd, adapter, QLCNIC_CMD_SET_NIC_INFO);
+ if (err)
+  return err;
+
+ cmd.req.arg[1] = QLC_FLOOD_MODE | QLC_VF_FLOOD_BIT;
+
+ err = qlcnic_issue_cmd(adapter, &cmd);
+ if (err)
+  dev_err(&adapter->pdev->dev,
+   "Failed to configure VF Flood bit on PF, err=%d\n",
+   err);
+
+ qlcnic_free_mbx_args(&cmd);
+ return err;
+}
+
 static int qlcnic_sriov_pf_cfg_eswitch(struct qlcnic_adapter *adapter,
            u8 func, u8 enable)
 {
@@ -467,6 +491,12 @@ static int qlcnic_sriov_pf_init(struct qlcnic_adapter *adapter)
  if (err)
   return err;
 
+ if (qlcnic_84xx_check(adapter)) {
+  err = qlcnic_sriov_pf_cfg_flood(adapter);
+  if (err)
+   goto disable_vlan_filtering;
+ }
+
  err = qlcnic_sriov_pf_cfg_eswitch(adapter, func, 1);
  if (err)
   goto disable_vlan_filtering;
-- 
1.7.1